### PR TITLE
Support trimming trailing slashes via a mount tuneable to support CMPv2

### PR DESCRIPTION
**Pull request by sgmiller**
### Description

CMPv2 specifies that the cmp path may or may not end in a trailing slash. 
Vault's request processing has a requirement that they do not for POST.  To
make this work, add a tuneable to the mount that will trim slashes if
enabled.  All existing mounts will default to off, preserving the existing
behavior.  CMPv2 capable PKI mounts will enable this.

### TODO only if you're a HashiCorp employee
- [ ] **Backport Labels:** If this PR is in the ENT repo and needs to be backported, backport  
  to N, N-1, and N-2, using the `backport/ent/x.x.x+ent` labels. If this PR is in the CE repo, you should only backport to N, using the `backport/x.x.x` label, not the enterprise labels.
    - [ ] If this fixes a critical security vulnerability or [severity 1](https://www.hashicorp.com/customer-success/enterprise-support) bug, it will also need to be backported to the current [LTS versions](https://developer.hashicorp.com/vault/docs/enterprise/lts#why-is-there-a-risk-to-updating-to-a-non-lts-vault-enterprise-version) of Vault. To ensure this, use **all** available enterprise labels.
- [ ] **ENT Breakage:** If this PR either 1) removes a public function OR 2) changes the signature
  of a public function, even if that change is in a CE file, _double check_ that
  applying the patch for this PR to the ENT repo and running tests doesn't
  break any tests. Sometimes ENT only tests rely on public functions in CE
  files.
- [ ] **Jira:** If this change has an associated Jira, it's referenced either
  in the PR description, commit message, or branch name.
- [ ] **RFC:** If this change has an associated RFC, please link it in the description.
- [ ] **ENT PR:** If this change has an associated ENT PR, please link it in the
  description. Also, make sure the changelog is in this PR, _not_ in your ENT PR.

### CODE REVIEW

File: vault/request_handling.go [hunk 1]
```diff
@@ -1007,6 +995,23 @@ func (c *Core) handleRequest(ctx context.Context, req *logical.Request) (retResp
 		if rawVals, ok := entry.synthesizedConfigCache.Load("audit_non_hmac_request_keys"); ok {
 			nonHMACReqDataKeys = rawVals.([]string)
 		}
+
+		// Allowing writing to a path ending in / makes it extremely difficult to
```
Comment (stevendpclark) [inline vault/request_handling.go:999]: Why did this move into handleRequest? I believe now we will do this check/error out quite a bit later in the process and this won't apply to login mounts (which is okay but still a little weird).

Comment (sgmiller) [reply vault/request_handling.go:999]: Oh, probably needs to be in both, but the reason is we don't have the mount entry in handleCancellableRequest.

File: vault/request_handling.go [hunk 2]
```diff
@@ -622,17 +627,18 @@ func (c *Core) handleCancelableRequest(ctx context.Context, req *logical.Request
 		(req.Operation == logical.UpdateOperation ||
 			req.Operation == logical.CreateOperation ||
 			req.Operation == logical.PatchOperation) {
-		return logical.ErrorResponse("cannot write to a path ending in '/'"), nil
+		if !entry.Config.TrimRequestTrailingSlashes {
```
Comment (stevendpclark) [inline vault/request_handling.go:630]:
```suggestion
		if entry == nil || !entry.Config.TrimRequestTrailingSlashes {
```

File: vault/router.go [hunk 3]
```diff
@@ -373,23 +373,28 @@ func (r *Router) MatchingMountByAccessor(mountAccessor string) *MountEntry {
...
 	path = ns.Path + path
 
-	mount, _, ok := r.root.LongestPrefix(path)
+	mount, raw, ok := r.root.LongestPrefix(path)
 	if !ok {
-		return ""
+		return "", nil
 	}
-	return mount
+	return mount, raw.(*MountEntry)
```
Comment (stevendpclark) [inline vault/router.go:397]:
```suggestion
	return mount, raw.(*routeEntry).mountEntry
```

File: vault/request_handling.go [hunk 4]
```diff
@@ -1007,6 +1012,23 @@ func (c *Core) handleRequest(ctx context.Context, req *logical.Request) (retResp
 		if rawVals, ok := entry.synthesizedConfigCache.Load("audit_non_hmac_request_keys"); ok {
 			nonHMACReqDataKeys = rawVals.([]string)
 		}
+
```
Comment (stevendpclark) [inline vault/request_handling.go:1015]: We should be able to get rid of this now?

File: vault/logical_system.go [hunk 5]
```diff
@@ -2753,6 +2753,58 @@ func (b *SystemBackend) handleTuneWriteCommon(ctx context.Context, path string,
...
+		}
+
+		mountEntry.SyncCache()
+
+		if b.Core.logger.IsInfo() {
+			b.Core.logger.Info("mount tuning of delegated_auth_accessors successful", "path", path)
+		}
+	}
+
+	if rawVal, ok := data.GetOk("trim_request_trailing_slashes"); ok {
```
Comment (stevendpclark) [inline vault/logical_system.go:2782]: Duplicated if here.

File: command/auth_tune.go [hunk 6]
```diff
@@ -306,6 +319,14 @@ func (c *AuthTuneCommand) Run(args []string) int {
 		if fl.Name == flagNameIdentityTokenKey {
 			mountConfigInput.IdentityTokenKey = c.flagIdentityTokenKey
 		}
+
+		if fl.Name == flagNameTrimRequestTrailingSlashes {
+			mountConfigInput.TrimRequestTrailingSlashes = c.flagTrimRequestTrailingSlashes
+		}
+
+		if fl.Name == flagNameTrimRequestTrailingSlashes {
```
Comment (stevendpclark) [inline command/auth_tune.go:327]: Duplicated if fl.Name..

Comment (sgmiller) [reply command/auth_tune.go:327]: Yeah made a bunch of mistakes using git diff + patch to bring changes over and having them conflict.

**Review by stevendpclark — APPROVED** [review]

**Pull request merged**